// ngx-onsenui/directives/ons-range.js
import { Directive, ElementRef, Input, Output, HostListener, EventEmitter, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * @element ons-range
 * @directive OnsRange
 * @selector ons-range
 * @description
 *   [en]Angular directive for `<ons-range>` component.[/en]
 *   [ja]`<ons-range>`要素のAngularディレクティブです。[/ja]
 * @example
 *   <ons-range [(value)]="foo"></ons-range>
 */
var OnsRange = (function () {
    function OnsRange(_elementRef) {
        this._elementRef = _elementRef;
        this._propagateChange = function (_) { };
        this._propagateTouched = function () { };
        /**
         * @output valueChange
         * @type {string}
         * @desc
         *   [en]Triggers when the value is changed.[/en]
         *   [ja]値が変更された時に発火します。[/ja]
         */
        this._valueChange = new EventEmitter();
        this._boundOnChange = this._onChange.bind(this);
        this._element = _elementRef.nativeElement;
        this._element.addEventListener('input', this._boundOnChange);
    }
    OnsRange.prototype._onChange = function (event) {
        this._valueChange.emit(this._element.value);
        this._propagateChange(this._element.value);
    };
    OnsRange.prototype._onBlur = function () {
        this._propagateTouched();
    };
    OnsRange.prototype.ngOnChanges = function (changeRecord) {
        var value = changeRecord['_value'].currentValue;
        this._element.value = value;
    };
    Object.defineProperty(OnsRange.prototype, "element", {
        get: function () {
            return this._element;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OnsRange.prototype, "nativeElement", {
        get: function () {
            return this._element;
        },
        enumerable: true,
        configurable: true
    });
    OnsRange.prototype.ngOnDestroy = function () {
        this._element.removeEventListener('input', this._boundOnChange);
        this._element = null;
    };
    OnsRange.prototype.writeValue = function (obj) {
        this._element.value = obj;
    };
    OnsRange.prototype.registerOnChange = function (fn) {
        this._propagateChange = fn;
    };
    OnsRange.prototype.registerOnTouched = function (fn) {
        this._propagateTouched = fn;
    };
    return OnsRange;
}());
export { OnsRange };
OnsRange.decorators = [
    { type: Directive, args: [{
                selector: 'ons-range',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(function () { return OnsRange; }),
                        multi: true,
                    }
                ]
            },] },
];
/** @nocollapse */
OnsRange.ctorParameters = function () { return [
    { type: ElementRef, },
]; };
OnsRange.propDecorators = {
    '_value': [{ type: Input, args: ['value',] },],
    '_valueChange': [{ type: Output, args: ['valueChange',] },],
    '_onBlur': [{ type: HostListener, args: ['blur',] },],
};
