// ngx-onsenui/directives/ons-search-input.js
import { Directive, ElementRef, Input, Output, HostListener, EventEmitter, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * @element ons-search-input
 * @directive OnsSearchInput
 * @selector ons-search-input
 * @description
 *   [en]Angular directive for `<ons-search-input>` component.[/en]
 *   [ja]`<ons-search-input>`要素のAngularディレクティブです。[/ja]
 * @example
 *   <ons-search-input [(value)]="value"></ons-search-input>
 */
var OnsSearchInput = (function () {
    function OnsSearchInput(_elementRef) {
        this._elementRef = _elementRef;
        this._propagateChange = function (_) { };
        this._propagateTouched = function () { };
        /**
         * @output valueChange
         * @type {string}
         * @desc
         *   [en]Triggers when the value is changed.[/en]
         *   [ja]内部の`input`要素の値が変更された時に発火します。[/ja]
         */
        this._valueChange = new EventEmitter();
        this._boundOnChange = this._onChange.bind(this);
        this._element = _elementRef.nativeElement;
        this._element.addEventListener('input', this._boundOnChange);
    }
    OnsSearchInput.prototype._onBlur = function () {
        this._propagateTouched();
    };
    OnsSearchInput.prototype._onChange = function (event) {
        this._valueChange.emit(this._element.value);
        this._propagateChange(this._element.value);
    };
    OnsSearchInput.prototype.ngOnChanges = function (changeRecord) {
        var value = changeRecord['_value'].currentValue;
        if (this._element.value !== value) {
            this._element.value = value;
        }
    };
    Object.defineProperty(OnsSearchInput.prototype, "element", {
        get: function () {
            return this._element;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OnsSearchInput.prototype, "nativeElement", {
        get: function () {
            return this._element;
        },
        enumerable: true,
        configurable: true
    });
    OnsSearchInput.prototype.ngOnDestroy = function () {
        this._element.removeEventListener('input', this._boundOnChange);
        this._element = null;
    };
    OnsSearchInput.prototype.writeValue = function (obj) {
        this._element.value = obj;
    };
    OnsSearchInput.prototype.registerOnChange = function (fn) {
        this._propagateChange = fn;
    };
    OnsSearchInput.prototype.registerOnTouched = function (fn) {
        this._propagateTouched = fn;
    };
    return OnsSearchInput;
}());
export { OnsSearchInput };
OnsSearchInput.decorators = [
    { type: Directive, args: [{
                selector: 'ons-search-input',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(function () { return OnsSearchInput; }),
                        multi: true,
                    }
                ]
            },] },
];
/** @nocollapse */
OnsSearchInput.ctorParameters = function () { return [
    { type: ElementRef, },
]; };
OnsSearchInput.propDecorators = {
    '_value': [{ type: Input, args: ['value',] },],
    '_valueChange': [{ type: Output, args: ['valueChange',] },],
    '_onBlur': [{ type: HostListener, args: ['blur',] },],
};
