// ngx-onsenui/directives/ons-input.js
import { Directive, ElementRef, Input, Output, HostListener, EventEmitter, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * @element ons-input
 * @directive OnsInput
 * @selector ons-input
 * @description
 *   [en]Angular directive for `<ons-input>` component.[/en]
 *   [ja]`<ons-input>`要素のAngularディレクティブです。[/ja]
 * @example
 *   <ons-input [(value)]="value"></ons-input>
 */
var OnsInput = (function () {
    function OnsInput(_elementRef) {
        this._elementRef = _elementRef;
        this._propagateChange = function (_) { };
        this._propagateTouched = function () { };
        /**
         * @output valueChange
         * @type {string}
         * @desc
         *   [en]Triggers when the value is changed.[/en]
         *   [ja]内部の`input`要素の値が変更された時に発火します。[/ja]
         */
        this._valueChange = new EventEmitter();
        this._boundOnChange = this._onChange.bind(this);
        this._element = _elementRef.nativeElement;
        this._element.addEventListener('input', this._boundOnChange);
    }
    OnsInput.prototype._onChange = function (event) {
        this._valueChange.emit(this._element.value);
        this._propagateChange(this._element.value);
    };
    OnsInput.prototype._onBlur = function () {
        this._propagateTouched();
    };
    OnsInput.prototype.ngOnChanges = function (changeRecord) {
        var value = changeRecord['_value'].currentValue;
        if (this._element.value !== value) {
            this._element.value = value;
        }
    };
    Object.defineProperty(OnsInput.prototype, "element", {
        get: function () {
            return this._element;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OnsInput.prototype, "nativeElement", {
        get: function () {
            return this._element;
        },
        enumerable: true,
        configurable: true
    });
    OnsInput.prototype.ngOnDestroy = function () {
        this._element.removeEventListener('input', this._boundOnChange);
        this._element = null;
    };
    OnsInput.prototype.writeValue = function (obj) {
        this._element.value = obj;
    };
    OnsInput.prototype.registerOnChange = function (fn) {
        this._propagateChange = fn;
    };
    OnsInput.prototype.registerOnTouched = function (fn) {
        this._propagateTouched = fn;
    };
    return OnsInput;
}());
export { OnsInput };
OnsInput.decorators = [
    { type: Directive, args: [{
                selector: 'ons-input',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(function () { return OnsInput; }),
                        multi: true,
                    }
                ]
            },] },
];
/** @nocollapse */
OnsInput.ctorParameters = function () { return [
    { type: ElementRef, },
]; };
OnsInput.propDecorators = {
    '_value': [{ type: Input, args: ['value',] },],
    '_valueChange': [{ type: Output, args: ['valueChange',] },],
    '_onBlur': [{ type: HostListener, args: ['blur',] },],
};
